### PR TITLE
Allow linalg.lstsq to use svd to compute the result for rank deficient matrices. 

### LINKED ISSUE #117122: Improve behaviour of `torch.linalg.lstsq` on CUDA GPU for rank defficient matrices
### 🚀 The feature, motivation and pitch

As per the documentation:
> For CUDA input [`torch.linalg.lstsq`] assumes that A is full-rank.

While documented, this behaviour is counter-intuitive for end-users especially if the function silently fails.

Interestingly, currently calling `torch.linalg.lstsq` on CUDA for rank defficient input silently fails in non-batched-mode but throws a `_LinAlgError` in batched mode.

It is also counter-intuitive that `torch.linalg.lstsq` on CUDA is not able to fallback to a more stable SVD driver despite `torch.linalg.svd` being supported on CUDA.

It would be great if:
1. The QR based implementation always threw an error is the input is not full rank
2. An SVD backend would be available on CUDA as well

### Alternatives

An alternative is to implement an SVD-based least-squares and use that instead of `torch.linalg.lstsq`. Here is a basic implementation (feel free to post refinements):
```python
def svd_lstsq(AA, BB, tol=1e-5):
    U, S, Vh = torch.linalg.svd(AA, full_matrices=False)
    Spinv = torch.zeros_like(S)
    Spinv[S>tol] = 1/S[S>tol]
    UhBB = U.adjoint() @ BB
    if Spinv.ndim!=UhBB.ndim:
      Spinv = Spinv.unsqueeze(-1)
    SpinvUhBB = Spinv * UhBB
    return Vh.adjoint() @ SpinvUhBB
```

### Additional context

Simple script to replicate:
```python
import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevice = torch.device('cpu')
if torch.cuda.is_available():
  torchdevice = torch.device('cuda')
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))
print('Running on ' + str(torchdevice))

b = 2
r = 5
c = 3
k = 1

if b==1:
  A = torch.randn(r, c, device=torchdevice)
  if k==1:
    B = torch.randn(r, device=torchdevice)
  else:
    B = torch.randn(r, k, device=torchdevice)
else:
  A = torch.randn(b, r, c, device=torchdevice)
  B = torch.randn(b, r, k, device=torchdevice)

# degrade rank
A[...,-1] = 0
print("A",A)

try:
  X_lstsq = torch.linalg.lstsq(A, B).solution
  print("X_lstsq",X_lstsq)
except Exception as error:
  print("An error occurred:", type(error).__name__, "–", error)

X_pinv=torch.linalg.pinv(A) @ B
print("X_pinv",X_pinv)

def svd_lstsq(AA, BB, tol=1e-5):
    U, S, Vh = torch.linalg.svd(AA, full_matrices=False)
    Spinv = torch.zeros_like(S)
    Spinv[S>tol] = 1/S[S>tol]
    UhBB = U.adjoint() @ BB
    if Spinv.ndim!=UhBB.ndim:
      Spinv = Spinv.unsqueeze(-1)
    SpinvUhBB = Spinv * UhBB
    return Vh.adjoint() @ SpinvUhBB

X_svd= svd_lstsq(A, B)
print("X_svd",X_svd)
```

Related issues: #88101 #85021 #10454

### PR DESCRIPTION
**Pull request by ZelboK**
Fixes #117122

This PR adds the logic so that in the case of rank deficient matrices, it can fallback to an SVD backend for batched mode. A big thank you to @tvercaut for the well written issue and suggestion on how to approach the problem. 

Summary:
1) At the time of writing this I haven't touched non-batched yet. I am hoping to get some feedback before proceeding. 
2) I believe there should be eyes on how specifically we want to fallback to SVD as the implementation when we run into rank deficient matrices. 

Please keep in mind this is my 2nd PR to pytorch, and I've never really used pytorch. I'm learning independently through digging deep in the internals so I may make some obvious mistakes. Please forgive!

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -3422,16 +3427,40 @@ static void linalg_lstsq_out_info(
...
   }
-
   // 'input' is modified in-place so we need a column-major copy
   auto input_working_copy = copyBatchedColumnMajor(input);
-
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if(driver == "gelsd" || driver == "gelss") {
+    auto k = std::min(m, n);
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3434]: So the way it's setup now is that if a user specifies `gelsd` or `gelss` as their driver(both use SVD) then it'll go through the SVD implementation. With that being said,

1) This isn't a fallback mechanism. This just merely provides the user the ability to do a lstsq on rank deficient matrices with SVD. 
I do not believe this is ideal. If we want to use a fallback mechanism, then ideally, we would check `input` or `A` tensor to see if it is rank deficient. However, I do not know how to do this in a performant way, so hopefully someone can point me in the right direction. 

If you do `lstsq_stub`, errors will be reflected in `_linalg_check_errors` after the computation. I would like to avoid having to do this and figure out from the jump so we can avoid the redundant work.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -3422,16 +3427,40 @@ static void linalg_lstsq_out_info(
...
+    auto Vh = at::zeros({input.size(0), k, n}, input.options());
+
+    svd_stub(input.device().type(),
+          input,
+          false, // you don't want the full SVD for least squares
+          true, // we need U, S, Vh for least squares
+          "gesvd",
+          U, singular_values, Vh, infos);
+
+    auto tol = 1e-5; // what should this be? can rcond be used?
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3445]: Like the comment says, please lmk if there is a suitable tolerance to be used.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -3422,16 +3427,40 @@ static void linalg_lstsq_out_info(
...
+  if(driver == "gelsd" || driver == "gelss") {
+    auto k = std::min(m, n);
+    auto U = at::zeros({input.size(0), m, k}, input.options());
+    auto Vh = at::zeros({input.size(0), k, n}, input.options());
+
+    svd_stub(input.device().type(),
+          input,
+          false, // you don't want the full SVD for least squares
+          true, // we need U, S, Vh for least squares
+          "gesvd",
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3442]: Not sure if the docs indirectly imply `gesvd` is "safer" than `gesvdj` because it says it'll fallabck to this . So for brevity, I'll leave it as this as a default. https://pytorch.org/docs/stable/generated/torch.linalg.svd.html

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -3490,8 +3520,8 @@ static std::string get_default_lstsq_driver(c10::optional<c10::string_view> driv
       );
     } else { // else if (input.is_cuda())
       TORCH_CHECK(
-        driver_str == "gels",
-        "torch.linalg.lstsq: `driver` other than `gels` is not supported on CUDA"
+        (driver_str == "gels" || driver_str == "gelsd"),
+        "torch.linalg.lstsq: `driver` other than `gels` or `gelsd` is not supported on CUDA"
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3524]: with the if case
`if (driver == "gels" || driver == "gelsd")` i added this so it would pass the assertion.

**Comment by ZelboK** [comment]
@janeyx99 @ptrblck 

2nd PR! Sorry it took me some days. I'm learning Pytorch internals independently so I'm still learning the codebase.

**Comment by lezcano** [comment]
mind cleaning up all the spurious new lines and the PR in general?

**Comment by ZelboK** [comment]
@lezcano 

My apologies! I've cleaned it up. I missed some new lines from when I was cleaning up my debugging/experimenting code.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (tvercaut) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3444]: In `numpy`, they now use the following:
> machine precision times ``max(M, N)`` where M and N are the input matrix dimensions
https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3433]: It's not possible to do detect this by default, that's why we ask the user to define the driver themselves.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -3427,8 +3429,33 @@ static void linalg_lstsq_out_info(
   auto input_working_copy = copyBatchedColumnMajor(input);
 
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if(driver == "gelsd" || driver == "gelss") {
+    auto k = std::min(m, n);
+    auto U = at::zeros({input.size(0), m, k}, input.options());
+    auto Vh = at::zeros({input.size(0), k, n}, input.options());
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3435]: zeros actually zeros out `U` and `Vh`. We don't need that. Rather than going so low level, you can call the higher level `at::_linalg_svd`.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3441]: `gesvd` is the right one to use here.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -3427,8 +3429,33 @@ static void linalg_lstsq_out_info(
   auto input_working_copy = copyBatchedColumnMajor(input);
 
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if(driver == "gelsd" || driver == "gelss") {
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3432]: This doesn't do any tridiagonal work, so we should just use it when `gelss` is chosen.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3444]: You should use `rcond` here.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 7]
```diff
@@ -3427,8 +3429,33 @@ static void linalg_lstsq_out_info(
...
+          false, // you don't want the full SVD for least squares
+          true, // we need U, S, Vh for least squares
+          "gesvd",
+          U, singular_values, Vh, infos);
+
+    auto tol = 1e-5; // what should this be? can rcond be used?
+    auto mask = singular_values > tol;
+    auto pseudo_sv = at::zeros_like(singular_values);
+
+    pseudo_sv.masked_scatter_(mask, singular_values.masked_select(mask).reciprocal());
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3448]:
```suggestion
    auto pseudo_sv = singular_values.reciprocal();
    pseudo_sv.masked_fill_(singular_values <= rcond, 0);
```
as per th edocs.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 4 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3521]: should be gelss.

**Review by lezcano — COMMENTED** [review]
It looks mostly good.

Needs tests in `test_linalg.py` and updating the docs noting that this `gelss` mode is also supported.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 6 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3432]: In that case, I'll also update this line ~3523
`(driver_str == "gelss" || driver_str == "gelssd"),`
to
`driver_str == "gelss"`

File: torch/linalg/__init__.py [hunk 8]
```diff
@@ -1032,7 +1032,7 @@
   - `'gelsd'` (tridiagonal reduction and SVD)
   - But if you run into memory issues: `'gelss'` (full SVD).
 
-For CUDA input, the only valid driver is `'gels'`, which assumes that :attr:`A` is full-rank.
+For CUDA inputs, two drivers are available: 'gels' and 'gelss'. The 'gels' driver assumes that :attr:A is full-rank. For rank-deficient matrices :attr:A, use the 'gelss' driver.
```
Comment (lezcano) [inline torch/linalg/__init__.py:1035]:
```suggestion
For CUDA inputs, two drivers are available: 'gels' and 'gelss'.
```
the drivers are already discussed in the previous paragraph.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 9]
```diff
@@ -3490,8 +3505,8 @@ static std::string get_default_lstsq_driver(c10::optional<c10::string_view> driv
       );
     } else { // else if (input.is_cuda())
       TORCH_CHECK(
-        driver_str == "gels",
-        "torch.linalg.lstsq: `driver` other than `gels` is not supported on CUDA"
+        driver_str == "gelss",
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3508]: This now breaks `gels`

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 10]
```diff
@@ -3490,8 +3505,8 @@ static std::string get_default_lstsq_driver(c10::optional<c10::string_view> driv
...
+        "torch.linalg.lstsq: `driver` other than `gels` or `gelsd` is not supported on CUDA"
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3509]: gelss

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 11]
```diff
@@ -3427,8 +3429,21 @@ static void linalg_lstsq_out_info(
   auto input_working_copy = copyBatchedColumnMajor(input);
 
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if(driver == "gelss") {
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3432]:
```suggestion
  if (driver == "gelss") {
```

**Comment by ZelboK** [comment]
@lezcano 

So when it comes to the tests, what kind of test did you think would be appropriate? Aside from checking that it no longer throws. I can add `gelss` as a driver to be used in `test_linalg_lstsq_batch_broadcasting` for example and assure its results are as expected.

File: test/test_linalg.py [hunk 12]
```diff
@@ -445,7 +445,15 @@ def complement_device(device):
...
-            with self.assertRaisesRegex(RuntimeError, '`driver` other than `gels` is not supported on CUDA'):
+            try:
+                result = torch.linalg.lstsq(a, b, driver='gelss')
+                self.assertTrue(result is not None)
+            except Exception as e:
+                self.fail(f"Unexpected error occurred: {e}")
+            with self.assertRaisesRegex(
+                RuntimeError,
+                'torch.linalg.lstsq: `driver` other than `gels` or `gelss` is not supported on CUDA'
+            ):
```
Comment (lezcano) [inline test/test_linalg.py:456]: I don't think this change is needed

**Review by lezcano — COMMENTED** [review]
For testing, just add a path that tests this driver in the relevant tests that tests the other drivers. We may even already have a test that tests this driver for CPU.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 13]
```diff
@@ -3450,7 +3456,7 @@ static void linalg_lstsq_out_info(
     bool compute_residuals = true;
     if (driver == "gelss" || driver == "gelsd") {
       if (input.dim() == 2) {
-        compute_residuals = (rank.item().toInt() == n);
+        compute_residuals = (rank.item().toDouble() == n);
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3459]: @lezcano 

So there is a problem where this would run into an overflow if left as `toInt()`. The reason is because rank will have non-sensical values if it goes through the SVD code path, since unlike `lstsq_kernel`, `svd_kernel` doesn't touch the rank tensors. Now with that being said, I don't like this as a work around. I'm hoping to get your input on how we should approach this situation.

I don't want to rework the entire function and add more cognitive load on you and other reviewers. Not without your opinion anyway.

File: test/test_linalg.py [hunk 14]
```diff
@@ -261,12 +261,13 @@ def select_if_not_empty(t, i):
 
         def check_correctness_scipy(a, b, res, driver, cond):
             # SciPy provides 3 driver options: gelsd, gelss, gelsy
-            if TEST_SCIPY and driver in ('gelsd', 'gelss', 'gelsy'):
-                import scipy.linalg
+            if device == 'cpu':
+                if TEST_SCIPY and driver in ('gelsd', 'gelss', 'gelsy'):
```
Comment (ZelboK) [inline test/test_linalg.py:265]: I was running into this error
```
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

And as I write this comment I should just transfer it to the host.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 15]
```diff
@@ -3427,15 +3429,34 @@ static void linalg_lstsq_out_info(
   auto input_working_copy = copyBatchedColumnMajor(input);
 
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if (driver == "gelss" && input.device() != at::kCPU) {
+    if (input.numel() == 0 || input.size(-2) == 0 || input.size(-1) == 0) {
+        auto output_shape = input.sizes().vec();
+        output_shape.back() = other.size(-1);
+        solution = at::zeros(output_shape, input.options());
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3436]: Is this necessary?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 13 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3459]: You can get the rank from the computation, but that would it would synchronise. Now, this is already synchronising, so I guess it's alright.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 16]
```diff
@@ -3444,8 +3465,9 @@ static void linalg_lstsq_out_info(
         compute_residuals = at::all(rank == n).item().toBool();
       }
     }
-    if (compute_residuals) {
-      // LAPACK stores residuals data for postprocessing in rows n:(m-n)
+  if (compute_residuals) {
+    // LAPACK stores residuals data for postprocessing in rows n:(m-n)
+    if (solution.size(-2) >= n + (m - n)) {
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3470]: why is all this necessary?

**Review by lezcano — COMMENTED** [review]
Please try to keep the changes to the bare minimum

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 15 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Sorry I should've communicated these changes.

The tests will actually fail without this check because it'll generate tensors similar to this `torch.empty((0, 1))`. The `narrow` code will lead to

```
  File "/home/ksm/pytorch/test/test_linalg.py", line 316, in test_linalg_lstsq
    res = torch.linalg.lstsq(a, b, rcond=rcond, driver=driver)
RuntimeError: start (0) + length (1) exceeds dimension size (0).
``` 

Do we want to remove the edge case handling to simplify the logic, and communicate in the docs that this will occur?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 16 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3473]: The tests were exposing some problems and they mainly originated from use of `narrow` in specific tensor cases. So these new conditionals were added to handle those situations. 

Errors like so: 

```
  File "/home/ksm/pytorch/test/test_linalg.py", line 316, in test_linalg_lstsq
    res = torch.linalg.lstsq(a, b, rcond=rcond, driver=driver)
RuntimeError: start (1) + length (1) exceeds dimension size (1)
```

Would it be preferable to revert these changes, run the test workflows, see what fails and go from there?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 17]
```diff
@@ -3427,8 +3429,30 @@ static void linalg_lstsq_out_info(
   auto input_working_copy = copyBatchedColumnMajor(input);
 
   // now the actual call that computes the result in-place (apply_lstsq)
-  lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
-
+  if (driver == "gelss" && input.device() != at::kCPU) {
+    if (input.numel() == 0 || input.size(-2) == 0 || input.size(-1) == 0) {
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3433]:
```suggestion
    if (input.numel() == 0) {
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 18]
```diff
@@ -3427,8 +3429,30 @@ static void linalg_lstsq_out_info(
...
+        auto output_shape = input.sizes().vec();
+        output_shape.back() = other.size(-1);
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3435]: This is not correct. But even more, `solution` has already been allocated, you just need to zero it out (for the case where it may not be empty)

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 16 above)
Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3473]: But this path was already working before. I don't understand why should we touch it at all?

File: test/test_linalg.py [hunk 19]
```diff
@@ -219,8 +219,8 @@ def select_if_not_empty(t, i):
             if a.numel() > 0:
                 for i in range(batch_size):
                     sol, residuals, rank, singular_values = ref(
-                        a_3d.select(0, i).numpy(),
-                        b_3d.select(0, i).numpy()
+                        a_3d.select(0, i).cpu().numpy(),
+                        b_3d.select(0, i).cpu().numpy()
```
Comment (lezcano) [inline test/test_linalg.py:223]: Is this change necessary?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 20]
```diff
@@ -3427,8 +3429,30 @@ static void linalg_lstsq_out_info(
...
+        rank.fill_(0);
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3436]: Rank is an empty tensor, this is not necessary.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 16 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3473]: It's not needed. When adding tests and trying to go through exceptions, I found `narrow` was the main cause. So I added guards against all the `narrow`s in the code. 
I added this because if the second last dimension was less than `n + m - n` (which I now realize is just `m`) then this will throw. 

I tried to produce a scenario where this would throw an exception, but it's always caught earlier on. This check is redundant and can be removed.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 21]
```diff
@@ -3456,13 +3481,19 @@ static void linalg_lstsq_out_info(
       at::sum_out(residuals, raw_residuals, /*dim=*/-2, /*keepdim=*/false, /*dtype*/real_dtype);
     }
   }
-  auto solution_view = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
-  // manually restride original
-  solution.set_(solution.storage(), solution_view.storage_offset(), solution_view.sizes(), solution_view.strides());
+  }
+  if (solution.size(-2) >= n) {
+    auto solution_view = solution.narrow(/*dim=*/-2, /*start=*/0, /*length*/n);
```
Comment (ZelboK) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3486]: @lezcano 

I added this, the tests for `test_linalg_lstsq` was failing for :
` python test_linalg.py -k test_linalg_lstsq_cuda_float32`

with tensors like:

```
torch.Size([2, 1]) is a shape
a contents: tensor([[0.1540],
        [0.9887]], device='cuda:0')
```

File: test/test_linalg.py (hunk 19 above)
Comment (ZelboK) [reply test/test_linalg.py:223]:
```
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```
Is what I ran into with this trace

```
  File "/home/ksm/pytorch/test/test_linalg.py", line 323, in test_linalg_lstsq
    check_correctness_scipy(a, b, res, driver, rcond)
  File "/home/ksm/pytorch/test/test_linalg.py", line 269, in check_correctness_scipy
    check_correctness_ref(a, b, res, scipy_ref, driver=driver)
  File "/home/ksm/pytorch/test/test_linalg.py", line 222, in check_correctness_ref
    a_3d.select(0, i).numpy(),
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```
After adding `gelss` as a driver for the test.

**Comment by ZelboK** [comment]
@cpuhrsch @lezcano Could one of you please run the pipeline / tests workflow to see if perhaps these tests are flaky on my environment for whatever reason?

File: test/test_linalg.py (hunk 19 above)
Comment (lezcano) [reply test/test_linalg.py:223]: But why was it passing before for the other inputs and the other drivers and not now?

Comment (ZelboK) [reply test/test_linalg.py:223]: @lezcano Sorry for late reply, haven't been feeling too well. 

So this test is not failing for other drivers - it's only failing for SVD and GPU. The other driver on teh GPU, `gels` doesn't even run. 

Tangentially, I was told that I could comment out the `skipCUDAIfNoMagma` for this PR(I do not have magma built). Could you please confirm that is ok? I should probably also look into devcontainers for managing dependencies when contributing.

Comment (lezcano) [reply test/test_linalg.py:223]: Oh, now I see the skip down below. Oh well.

File: torch/linalg/__init__.py [hunk 22]
```diff
@@ -1080,7 +1080,7 @@
 
 Keyword args:
     driver (str, optional): name of the LAPACK/MAGMA method to be used.
-        If `None`, `'gelsy'` is used for CPU inputs and `'gels'` for CUDA inputs.
+        If `None`, `'gelsy'` is used for CPU inputs, `'gels'` and `'gelss'` for CUDA inputs.
```
Comment (lezcano) [inline torch/linalg/__init__.py:1083]: revert

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 23]
```diff
@@ -3427,15 +3429,36 @@ static void linalg_lstsq_out_info(
...
+    lstsq_stub(input.device().type(), input_working_copy, solution, rank, singular_values, infos, rcond, driver);
+  }
   // residuals are available only if m > n and drivers other than gelsy used
   if (m > n && driver != "gelsy") {
     // if the driver is gelss or gelsd then the residuals are available only if rank == n
     bool compute_residuals = true;
     if (driver == "gelss" || driver == "gelsd") {
       if (input.dim() == 2) {
-        compute_residuals = (rank.item().toInt() == n);
+        compute_residuals = (rank.item().toDouble() == n);
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3461]: why is this change necessary?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 24]
```diff
@@ -3427,15 +3429,36 @@ static void linalg_lstsq_out_info(
...
-
+  if (driver == "gelss" && input.device() != at::kCPU) {
+    if (input.numel() == 0) {
+        auto output_shape = input.sizes().vec();
+        output_shape.back() = other.size(-1);
+        solution.zero_();
+    } else {
+        auto [U, S, Vh] = at::_linalg_svd(input, false, true, "gesvd");
+        rank = at::zeros({1}, at::kLong);
+        rank[0] = (S > rcond).sum();
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3440]: This is not correct. Compute the rank by looking at the zeros of `S_pinv`.

**Review by lezcano — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 23 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3461]: Before it would run into integer overflows when the rank wasn't computed. Since we are now computing the rank this can be changed back.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 24 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3440]: For my understanding(hopefully not a stupid question) - is it because the way it's done currently is unstable or it leads to inaccurate results?

Comment (lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3440]: It's not about being unstable. It's about the way we define what is a zero singular value in this function.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 25]
```diff
@@ -3427,8 +3429,30 @@ static void linalg_lstsq_out_info(
...
+        output_shape.back() = other.size(-1);
+        solution.zero_();
+    } else {
+        auto [U, S, Vh] = at::_linalg_svd(input, false, true, "gesvd");
+        rank = at::zeros({1}, at::kLong);
+
+        auto S_pinv = S.reciprocal();
+        auto s1 = at::narrow(S, /*dim=*/-1, /*start=*/0, /*length=*/1);  // singular values are sorted in descending order
+        S_pinv.masked_fill_(S < rcond * s1, 0);
+        rank[0] = (S != 0).sum();
```
Comment (lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3444]: This is not correct. Please see the definition of what is considered a zero singular value in the docs of this function.

**Review by lezcano — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 25 above)
Comment (ZelboK) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3444]: I meant to use the pseudo inverse here, not `S`, that is my bad.

**Comment by lezcano** [comment]
bad rebase. Too many people are tagged now. Open a new PR.
